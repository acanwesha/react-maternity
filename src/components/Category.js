import React, { useEffect, useState } from 'react'
import { Collapse, ListGroup, ListGroupItem } from "react-bootstrap"

const Category = props => {
  const {
    category,
    categoryIndex,
    selectedCategory,
    setSelectedCategory,
    methods,
    finalCategoryArray,
  } = props
  const { setCategoryArray } = methods

  const [collapsedState, setCollapsedState] = useState(
    categoryIndex === selectedCategory //  ===-1
  )
  const [subCheckbox, setSubCheckbox] = useState(false)

  useEffect(() => {
    setCollapsedState(categoryIndex === selectedCategory)
  }, [categoryIndex, selectedCategory])

  return (
    <div>
      <ListGroup>
        <span>
          <ListGroupItem style={{ backgroundColor: "#cceeff" }}>
            <div>
              <input
                type='checkbox'
                name='categoryId'
                data-name="category-checkbox"
                // id={category.id}
                onChange={e => {
                  setSubCheckbox(!subCheckbox)
                  const displayName = category.displayName
                  if (e.target.checked) {
                    //  if selected then added new category
                    const updatedArray = [
                      ...finalCategoryArray,
                      { displayName: displayName },
                    ]
                    setCategoryArray(updatedArray)
                    //                    console.log("added category", updatedArray)
                  } else {
                    //  if unselected then remove category from array
                    const updatedArray = finalCategoryArray.filter(
                      obj => obj.displayName !== displayName
                    )
                    setCategoryArray(updatedArray)
                    //                  console.log("removed category", updatedArray)
                  }
                }}
              />
              <label
                //    htmlFor={category.id}
                data-testid="displayName"
                style={{ cursor: "pointer" }}
                onClick={() => {
                  setSelectedCategory(categoryIndex)
                  setCollapsedState(!collapsedState)
                }}
                className='ml-2'
              >
                {category.displayName}
              </label>
            </div>
            <Collapse
              in={collapsedState}
              style={{
                position: "absolute",
                left: "348px",
                top: "0px",
                width: "450px",
                backgroundColor: "#dcf9fa"
              }}
            >
              <div id={category.id}>
                {category.result.map((item, index) => {
                  if (Object.keys(item).length === 0) {
                    return ""
                  }
                  //  if we unselect the main category then also need to unselect the sub categories
                  if (!subCheckbox) {
                    var checkboxElement = document.getElementById(item.id)
  
                    if (checkboxElement !== null) {
                      if (checkboxElement.checked === true) {
                        checkboxElement.checked = false
                      }
                    }
                  }
                  return (
                    <ListGroup 
                      key={index} 
                      className='ml-1'
                      data-name="item"
                    >
                      <span>
                        <ListGroupItem style={{ backgroundColor: "#caecfc" }}>
                          <div> 
                            <input
                              type='checkbox'
                              name='itemId'
                              data-name="sub-category-checkbox"
                              id={item.id}
                              disabled={!subCheckbox}
                              onChange={e => {
                                const displayName = item.displayName

                                //  first get the category index
                                const currentCategoryId =
                                  finalCategoryArray.findIndex(
                                    object =>
                                      object.displayName === category.displayName
                                  )
                                //  create a duplicate array for modification
                                const duplicateArray = [...finalCategoryArray]
                                var resultArray = []
                                if (
                                  duplicateArray[currentCategoryId].result !==
                                  undefined
                                ) {
                                  resultArray = [
                                    ...duplicateArray[currentCategoryId].result,
                                  ]
                                }
                                if (e.target.checked) {
                                  //  added into array
                                  const resultObject = {
                                    displayName: displayName,
                                  }
                                  if (item.value !== undefined) {
                                    resultObject["value"] = item.value
                                  }
                                  resultArray.push(resultObject)
                                  duplicateArray[currentCategoryId].result =
                                    resultArray
                                  setCategoryArray(duplicateArray)
                                   console.log(
                                    "sub-category added ",
                                    duplicateArray
                                  )
                                } else {
                                  // removed from array
                                  duplicateArray[currentCategoryId].result =
                                    resultArray.filter(
                                      object => object.displayName !== item.displayName
                                    )

                                  setCategoryArray(duplicateArray)
                                  console.log(
                                    "sub-category removed ",
                                    duplicateArray
                                  )
                                }
                              }}
                            />
                            <label
                              htmlFor={item.id}
                              style={{ cursor: "pointer" }}
                              className='ml-2'
                            >
                              {item.displayName}
                            </label>
                            {item.value !== undefined && (
                              <span>
                                : <b> {item.value}</b>
                              </span>
                            )}
                          </div>
                        </ListGroupItem>
                      </span>
                    </ListGroup>
                  )
                })}
              </div>
            </Collapse>
          </ListGroupItem>
        </span>
      </ListGroup>
    </div>
  )
}
export default Category;
