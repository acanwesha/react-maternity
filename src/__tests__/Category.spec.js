import React from 'react';
import { render, fireEvent } from '@testing-library/react'
import Category from '../components/Category';


describe('components/Category', () => {
  let wrapper;
  let props;

  const setup = () => {
    wrapper = render(<Category {...props} />)
  }

  const updateProps = () => {
    render(<Category {...props} />, {
      container: wrapper.container
    })
  }

  const getDisplayNameLabel = () => wrapper.container.querySelector('[data-testid="displayName"]')

  beforeEach(() => {
    props = {
      category: {
        result: [
          {
            id: 1,
            displayName: "sub-category 1",
            value: 10
          }, {
            id: 2,
            displayName: "sub-category 2",
            value: 20
          }
        ],
        displayName: "Test"
      },
      categoryIndex: 0,
      setSelectedCategory: jest.fn(),
      selectedCategory: 0,
      finalCategoryArray: [],
      methods: {
        setCategoryArray: jest.fn(),
      }
    }
  })

  it('should renders correctly', () => {
    setup();
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should render correctly when item is empty object', () => {
    props.category.result[0] = {}
    setup();
    expect(wrapper.container).toMatchSnapshot();
  })

  it('should handle check/uncheck category correctly', () => {
    setup();

    //Check
    fireEvent.click(wrapper.container.querySelector('input[data-name="category-checkbox"]'))

    expect(props.methods.setCategoryArray).toHaveBeenCalledWith([{
      displayName: "Test",
    }]);
    // All sub-category checkboxes are not disabled
    wrapper.container.querySelectorAll('input[data-name="sub-category-checkbox"]').forEach(checkbox => expect(checkbox).not.toBeDisabled())

    // category is added
    props.finalCategoryArray = [
      {
        displayName: "Test",
        result: []
      }
    ]
    updateProps();

    //Uncheck
    fireEvent.click(wrapper.container.querySelector('input[data-name="category-checkbox"]'))
    expect(props.methods.setCategoryArray).toHaveBeenCalledWith([]);

    // All sub-category checkboxes are disabled
    wrapper.container.querySelectorAll('input[data-name="sub-category-checkbox"]').forEach(checkbox => expect(checkbox).toBeDisabled())
  })

  it('should collapse/expand category correctly in case current category is selected', async () => {
    setup();
    // the category is selected initially
    expect(document.querySelectorAll('.collapse.show')).toHaveLength(1)
    fireEvent.click(getDisplayNameLabel())
    await new Promise((r) => setTimeout(r, 300)); // wait for animation
    expect(document.querySelectorAll('.collapse.show')).toHaveLength(0)
    fireEvent.click(getDisplayNameLabel())
    await new Promise((r) => setTimeout(r, 300)); // wait for animation
    expect(document.querySelectorAll('.collapse.show')).toHaveLength(1)
  });

  it('should collapse if the category is not selected', () => {
    setup();
    // the category is selected initially
    expect(document.querySelectorAll('.collapse.show')).toHaveLength(1)
    props.selectedCategory = 1;
    updateProps();
    expect(document.querySelectorAll('.collapse.show')).toHaveLength(0)
  })

  it('should handle check/uncheck sub-category correctly', () => {
    // category is added
    props.finalCategoryArray = [
      {
        displayName: "Test",
        result: []
      }
    ]
    
    setup();
    fireEvent.click(wrapper.container.querySelector('input[data-name="category-checkbox"]'))

    //Check first sub-category
    fireEvent.click(wrapper.container.querySelector('input[data-name="sub-category-checkbox"]'));
    expect(props.methods.setCategoryArray).toHaveBeenCalledWith([{
      result: [
        {
          displayName: "sub-category 1",
          value: 10
        }
      ],
      displayName: "Test"
    }]);
    fireEvent.click(wrapper.container.querySelector('input[data-name="sub-category-checkbox"]'));
    wrapper.container.querySelector('input[data-name="sub-category-checkbox"]').checked = true
    expect(props.methods.setCategoryArray).toHaveBeenCalledWith([{
      result: [],
      displayName: "Test"
    }]);

    // uncheck category
    fireEvent.click(wrapper.container.querySelector('input[data-name="category-checkbox"]'))
    expect(wrapper.container.querySelector('input[data-name="sub-category-checkbox"]').checked).toBe(false)
  })
})