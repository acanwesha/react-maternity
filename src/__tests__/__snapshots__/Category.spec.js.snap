// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Category should render correctly when item is empty object 1`] = `
<div>
  <div>
    <div
      class="list-group"
    >
      <span>
        <div
          class="list-group-item"
          style="background-color: rgb(204, 238, 255);"
        >
          <div>
            <input
              data-name="category-checkbox"
              name="categoryId"
              type="checkbox"
            />
            <label
              class="ml-2"
              data-testid="displayName"
              style="cursor: pointer;"
            >
              Test
            </label>
          </div>
          <div
            class="collapse show"
            style="position: absolute; left: 348px; top: 0px; width: 450px; background-color: rgb(220, 249, 250);"
          >
            
            <div
              class="ml-1 list-group"
              data-name="item"
            >
              <span>
                <div
                  class="list-group-item"
                  style="background-color: rgb(202, 236, 252);"
                >
                  <div>
                    <input
                      data-name="sub-category-checkbox"
                      disabled=""
                      id="2"
                      name="itemId"
                      type="checkbox"
                    />
                    <label
                      class="ml-2"
                      for="2"
                      style="cursor: pointer;"
                    >
                      sub-category 2
                    </label>
                    <span>
                      : 
                      <b>
                         
                        20
                      </b>
                    </span>
                  </div>
                </div>
              </span>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
</div>
`;

exports[`components/Category should renders correctly 1`] = `
<div>
  <div>
    <div
      class="list-group"
    >
      <span>
        <div
          class="list-group-item"
          style="background-color: rgb(204, 238, 255);"
        >
          <div>
            <input
              data-name="category-checkbox"
              name="categoryId"
              type="checkbox"
            />
            <label
              class="ml-2"
              data-testid="displayName"
              style="cursor: pointer;"
            >
              Test
            </label>
          </div>
          <div
            class="collapse show"
            style="position: absolute; left: 348px; top: 0px; width: 450px; background-color: rgb(220, 249, 250);"
          >
            <div
              class="ml-1 list-group"
              data-name="item"
            >
              <span>
                <div
                  class="list-group-item"
                  style="background-color: rgb(202, 236, 252);"
                >
                  <div>
                    <input
                      data-name="sub-category-checkbox"
                      disabled=""
                      id="1"
                      name="itemId"
                      type="checkbox"
                    />
                    <label
                      class="ml-2"
                      for="1"
                      style="cursor: pointer;"
                    >
                      sub-category 1
                    </label>
                    <span>
                      : 
                      <b>
                         
                        10
                      </b>
                    </span>
                  </div>
                </div>
              </span>
            </div>
            <div
              class="ml-1 list-group"
              data-name="item"
            >
              <span>
                <div
                  class="list-group-item"
                  style="background-color: rgb(202, 236, 252);"
                >
                  <div>
                    <input
                      data-name="sub-category-checkbox"
                      disabled=""
                      id="2"
                      name="itemId"
                      type="checkbox"
                    />
                    <label
                      class="ml-2"
                      for="2"
                      style="cursor: pointer;"
                    >
                      sub-category 2
                    </label>
                    <span>
                      : 
                      <b>
                         
                        20
                      </b>
                    </span>
                  </div>
                </div>
              </span>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
</div>
`;
